src/Vault.sol

```diff
@@ -0,0 +1,398 @@
+// SPDX-License-Identifier: BSD-3-Clause
+pragma solidity ^0.8.24;
+
+import {
+    AccessControlUpgradeable,
+    Address,
+    ERC20PermitUpgradeable,
+    ERC20Upgradeable,
+    IERC20,
+    IERC20Metadata,
+    Math,
+    ReentrancyGuardUpgradeable,
+    SafeERC20
+} from "./Common.sol";
+
+import {IVault} from "src/interface/IVault.sol";
+import {IStrategy} from "src/interface/IStrategy.sol";
+import {IRateProvider} from "src/interface/IRateProvider.sol";
+
+contract Vault is IVault, ERC20PermitUpgradeable, AccessControlUpgradeable, ReentrancyGuardUpgradeable {
+    using SafeERC20 for IERC20;
+    using Address for address;
+    using Math for uint256;
+
+    function decimals() public view virtual override(ERC20Upgradeable, IERC20Metadata) returns (uint8) {
+        AssetStorage storage assetStorage = _getAssetStorage();
+        return assetStorage.assets[assetStorage.list[0]].decimals;
+    }
+
+    function asset() public view returns (address) {
+        return _getAssetStorage().list[0];
+    }
+
+    function totalAssets() public view returns (uint256) {
+        return _getVaultStorage().totalAssets;
+    }
+
+    function convertToShares(uint256 assets) public view returns (uint256) {
+        return _convertToShares(asset(), assets, Math.Rounding.Floor);
+    }
+
+    function convertToAssets(uint256 shares) public view returns (uint256) {
+        return _convertToAssets(asset(), shares, Math.Rounding.Floor);
+    }
+
+    function maxDeposit(address) public pure returns (uint256) {
+        return type(uint256).max;
+    }
+
+    function maxMint(address) public view returns (uint256) {
+        if (paused()) return 0;
+        return type(uint256).max;
+    }
+
+    function maxWithdraw(address owner) public view returns (uint256) {
+        if (paused()) return 0;
+
+        // Here we return whichever is less, the available base assets in the buffer, or the user's
+        // vault token balance converted to base asset value
+        uint256 ownerAssets = _convertToAssets(asset(), balanceOf(owner), Math.Rounding.Floor);
+        // The buffer strategy must have the same base underlying asset as the vault
+        uint256 idleBalance = IERC20(asset()).balanceOf(address(this));
```

Comment:
questioning here if we should use the freshly deposited funds to fulfil withdrawals.

It's more controlled if withdrawers only have access to the buffer funds, not to fresh deposits as well.

CC @gearcat0  do you have more opinions here on withdrawing funds that were just deposited

Reply:
I don't object for the vault

Reply:
I've removed taking withdraws from idleBalances because it doubled the gas cost for withdraws. Now all withdraws come from the buffer strategy and get send to the owner from the strategy vault.